xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererMediaCodecSurface.cpp

```diff
@@ -156,6 +156,17 @@ bool CRendererMediaCodecSurface::RenderUpdateVideoHook(bool clear, DWORD flags,
     mci->RenderUpdate(srcRect, dstRect);
   }
 
+  /** @todo This is a really bad hack and not meant to be
+  * a solution. This needs proper fixing when android maintainer
+  * is back. Real solution: Block until your next vblank.
```

Comment:
> 2) is "wait((1/fps) - (current_time - m_lastRenderTime))", right?

Depends on the amount of internal buffers used. If triple buffers are used, it should also take m_lastRenderTime-1 into account. Such a system can recover even it was already late more than frametime.

> 3) Where / how do you let player know? This is actually trivial, because you can assume it's the time of ::RenderUpdate() with an error of 1 frame at worst

Due to the fact that current architecture is outgrown, the main loop has to serve input, gui rendering, video rendering (by-pass). VideoPlayer assumes that when a new cycle starts and last rendered frame was frame(n), 
frame(n - (g_Windowing.NoOfBuffers -1)) has made it to display. (additional display latency was omitted here)
In other words: If internal buffers are full, RenderUpdate should block until frame(n - (g_Windowing.NoOfBuffers -1)) has become visible.
This is the behaviour of swapBuffers and by-pass renderers have to emulate this behaviour if they want to stay in sync with the rest of the system.